quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -0,0 +1,912 @@
+        // Division with numbers or π (π / π already covered)
+        (Expression::Number(x), InfixOperator::Slash, Expression::Number(y)) => {
+            Expression::Number(x / y)
+        }
+        (Expression::Number(x), InfixOperator::Slash, Expression::PiConstant) => {
+            Expression::Number(x / PI)
+        }
+        (Expression::PiConstant, InfixOperator::Slash, Expression::Number(y)) => {
+            Expression::Number(PI / y)
+        }
+
+        // Exponentiation
+
+        // Exponentiation with zero
+        (Expression::Number(x), InfixOperator::Caret, _) if is_zero(x) => Expression::Number(ZERO),
+        (_, InfixOperator::Caret, Expression::Number(y)) if is_zero(y) => Expression::Number(ONE),
+        // Exponentiation with one
+        (Expression::Number(x), InfixOperator::Caret, _) if is_one(x) => Expression::Number(ONE),
+        (left, InfixOperator::Caret, Expression::Number(y)) if is_one(y) => left,
+
+        //----------------------------------------------------------------
+        // Next: dealing with negation in subexpressions
+        //----------------------------------------------------------------
+
+        // Addition with negation
+        // a + (-b) = (-b) + a = a - b
+        (
+            ref other,
+            InfixOperator::Plus,
+            Expression::Prefix(PrefixExpression {
+                operator: PrefixOperator::Minus,
+                ref expression,
+            }),
+        )
+        | (
+            Expression::Prefix(PrefixExpression {
+                operator: PrefixOperator::Minus,
+                ref expression,
+            }),
+            InfixOperator::Plus,
+            ref other,
+        ) => simplify_infix(
+            other,
+            InfixOperator::Minus,
+            expression,
+            limit.saturating_sub(1),
+        ),
+
+        // Subtraction with negation
+
+        // a - (-b) = a + b
+        (
+            ref left,
+            InfixOperator::Minus,
+            Expression::Prefix(PrefixExpression {
+                operator: PrefixOperator::Minus,
+                ref expression,
+            }),
+        ) => simplify_infix(
+            left,
+            InfixOperator::Plus,
+            expression,
+            limit.saturating_sub(1),
+        ),
+
+        // -expression - right = smaller of [(-expression) - right, -(expression + right)]
+        (
+            ref left @ Expression::Prefix(PrefixExpression {
+                operator: PrefixOperator::Minus,
+                ref expression,
+            }),
+            InfixOperator::Minus,
+            ref right,
+        ) => {
+            let original = sub!(left.clone(), right.clone());
+            let new = simplify_prefix(
+                PrefixOperator::Minus,
+                &simplify_infix(
+                    expression,
+                    InfixOperator::Plus,
+                    right,
+                    limit.saturating_sub(2),
```

Comment:
Why limit -2 here?  Calling `simplify_infix` just adds one stack frame which will be popped before the subsequent call to `simplify_prefix`.

Reply:
I subtracted two since this was inside two layers of recursion, as it were. If we only subtract one, doesn’t this one case get to recurse deeper than others?

Reply:
I guess if you're counting depth in the expression rather than frames on the simplifier's call stack, you're right.

Reply:
Oh, of course we should only worry about the call stack, that's the whole point. Plus, that would alleviate the need for the slightly annoying `.saturating_sub` calls everywhere, because we already check for zero.